src/app/locator/qgsinbuiltlocatorfilters.cpp

```diff
@@ -364,15 +367,16 @@ void QgsAllLayersFeaturesLocatorFilter::prepare( const QString &string, const Qg
                                    string ) );
     req.setLimit( 30 );
 
-    PreparedLayer preparedLayer;
-    preparedLayer.expression = expression;
-    preparedLayer.context = context;
-    preparedLayer.layerId = layer->id();
-    preparedLayer.layerName = layer->name();
-    preparedLayer.iterator =  layer->getFeatures( req );
-    preparedLayer.layerIcon = QgsMapLayerModel::iconForLayer( layer );
+    PreparedLayer *preparedLayer = new PreparedLayer();
+    preparedLayer->expression = expression;
+    preparedLayer->context = context;
+    preparedLayer->layerId = layer->id();
+    preparedLayer->layerName = layer->name();
+    preparedLayer->featureSource.reset( new QgsVectorLayerFeatureSource( layer ) );
+    preparedLayer->request = req;
+    preparedLayer->layerIcon = QgsMapLayerModel::iconForLayer( layer );
 
-    mPreparedLayers.append( preparedLayer );
+    mPreparedLayers.append( std::shared_ptr<PreparedLayer>( preparedLayer ) );
```

Comment:
I would initialize that already on line 370, then memory management is exception proof straight from the beginning.